src/Client.php

```diff
@@ -105,33 +131,64 @@ public function runQuery($query, bool $resultsAsArray = false, array $variables
      */
     public function runRawQuery(string $queryString, $resultsAsArray = false, array $variables = []): Results
     {
-        $request = new Request($this->requestMethod, $this->endpointUrl);
-
-        foreach($this->httpHeaders as $header => $value) {
-            $request = $request->withHeader($header, $value);
-        }
+        $request = $this->requestFactory->createRequest($this->requestMethod, $this->endpointUrl);
 
         // Convert empty variables array to empty json object
         if (empty($variables)) $variables = (object) null;
+
         // Set query in the request body
         $bodyArray = ['query' => (string) $queryString, 'variables' => $variables];
-        $request = $request->withBody(Psr7\stream_for(json_encode($bodyArray)));
 
-        // Send api request and get response
+        $stream = $this->streamFactory->createStream(json_encode($bodyArray));
+
+        $request = (new RequestDecorator())->decorate($request, $stream, $this->httpOptions, $this->httpHeaders);
+
         try {
+            $request = $request->withBody($stream);
+
+            foreach($this->httpHeaders as $header => $value) {
+                $request = $request->withHeader($header, $value);
+            }
+
+            // Send api request and get response
             $response = $this->httpClient->sendRequest($request);
         }
-        catch (ClientException $exception) {
-            $response = $exception->getResponse();
+        catch (BaseInvalidArgumentException $e) {
+            throw new InvalidArgumentException($e->getMessage(), $e->getCode(), $e);
+        }
+        catch (RequestExceptionInterface $e) {
+            throw RequestException::create($e->getRequest(), null, $e);
+        }
+        catch (NetworkExceptionInterface $e) {
+            throw ConnectException::create($e->getRequest(), null, $e);
+        }
+        catch (ClientExceptionInterface $e) {
```

Comment:
While I understand why you're mapping the `ClientExceptionInterface` to the `RequestException`, I can see how this might potentially be confusing to somebody handling the exceptions. I do not expect you to change anything related to the implementation, just want to know if you have any concerns about this?

Reply:
I have concerns about catching these exceptions altogether. I'm wondering if it's a better idea to let client exceptions reach the user so that user can know what exceptions will be thrown based on the client they are using. However my idea with this was to throw exceptions that were similar to the exceptions Guzzle ^6.0 would throw. 

The `RequestException::create(...)` will actually throw either a `ClientException` or `ServerException` if there is a response based on the error code, or just throw a `RequestException` if there's no response. I copied it from Guzzle ^6.0.

The ClientExceptionInterface is thrown when...
> if and only if it is unable to send the HTTP request at all or if the HTTP response could not be parsed into a PSR-7 response object

I think in that case Guzzle 6 would throw a RequestException which is why I used that. But again I don't know if doing it that way is a good idea. What do you think?

Reply:
Let's merge this discussion with the one we're having separately about exception handling at large.